### PR TITLE
seastore: add blk_paddr_t

### PR DESCRIPTION
**Pull request by myoungwon**
https://github.com/ceph/ceph/pull/43795
Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [ ] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/seastore_types.h [hunk 1]
```diff
@@ -377,6 +377,162 @@ class segment_map_t {
...
+  constexpr block_off_t(device_block_off_t offset) : off(offset) {}
+  constexpr uint32_t get_raw_bits() const {
+    check_valid(off);
+    uint32_t ret = off & 0xFFFFF;
+    if (off < 0) {
+      ret |= 1 << BLK_OFF_LEN_BITS;
+    }
+    return ret;
+  }
+  bool operator==(const block_off_t& other) const {
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:493]: I don't think we need separate block_off_t and segment_off_t types -- how about we just rename segment_off_t seastore_off_t?

File: src/crimson/os/seastore/seastore_types.h [hunk 2]
```diff
@@ -581,6 +753,92 @@ struct seg_paddr_t : public paddr_t {
...
+    dev_addr |= ((uint32_t)off.get_raw_bits() >> 12);
+  }
+
+  paddr_t add_offset(block_off_t o) const {
+    return paddr_t::make_blk_paddr(get_block_id(), get_block_off() + o);
+  }
+
+  paddr_t add_relative(paddr_t o) const {
+    assert(o.is_relative());
+    blk_paddr_t& s = o.as_blk_paddr();
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:791]: I don't think we want separate representations for relative addresses between block and segment -- they are semantically identical.

**Comment by myoungwon** [comment]
@athanatos Can you take a look to check I address your comments correctly?

**Comment by athanatos** [comment]
This patch seems to add several types:

device_block_id_t -- if this is meant to be similar to device_segment_id_t, it would be just a block offset.  Is this just supposed to be an unsigned seastore_off_t?
block_id_t -- combination of device_id_t and device_block_id_t as with segment_id_t?  Is this actually useful?  We want a segment_id_t because segment_cleaner internally deals with segments from multiple devices, but I don't see an obvious analogue for block devices that wouldn't just be a paddr_t.
blk_paddr_t -- this makes sense

I don't think you actually need device_block_id_t or block_id_t.

File: src/crimson/os/seastore/seastore_types.h [hunk 3]
```diff
@@ -174,13 +174,14 @@ std::ostream &segment_to_stream(std::ostream &, const segment_id_t &t);
 
 // Offset within a segment on disk, see SegmentManager
 // may be negative for relative offsets
-using segment_off_t = int32_t;
-constexpr segment_off_t NULL_SEG_OFF =
-  std::numeric_limits<segment_off_t>::max();
-constexpr segment_off_t MAX_SEG_OFF =
-  std::numeric_limits<segment_off_t>::max();
+using seastore_off_t = int32_t;
+using device_seastore_off_t = uint32_t;
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:178]: What is a device_seastore_off_t?

File: src/crimson/os/seastore/seastore_types.h [hunk 4]
```diff
@@ -377,6 +378,104 @@ class segment_map_t {
...
+  std::numeric_limits<seastore_off_t>::min() /
+  (1 << std::numeric_limits<seastore_off_t>::digits - BLK_OFF_LEN_BITS);
+constexpr uint32_t BLK_OFF_NUM_MASK = 0x7FFFF;
+constexpr uint32_t BLK_OFF_MSB_MASK = 0x80000;
+constexpr seastore_off_t NULL_BLK_OFF = BLK_OFF_MAX;
+constexpr seastore_off_t MAX_BLK_OFF = BLK_OFF_MAX;
+static void check_blk_off_valid(const seastore_off_t offset) {
+  assert(offset < BLK_OFF_MAX);
+  assert(offset > BLK_OFF_MIN);
+}
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:477]: Why do we need both MAX_BLK_OFF and BLK_OFF_MAX?  Rename NULL_BLK_OFF to BLK_OFF_NULL.

File: src/crimson/os/seastore/seastore_types.h [hunk 5]
```diff
@@ -377,6 +378,104 @@ class segment_map_t {
...
+
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:478]: Also, I think these are all implementation details of blk_paddr_t, can we move these into the private: section of blk_paddr_t?

File: src/crimson/os/seastore/seastore_types.h [hunk 6]
```diff
@@ -401,35 +500,54 @@ enum class addr_types_t : uint8_t {
...
 protected:
   using common_addr_t = uint64_t;
   common_addr_t dev_addr;
 private:
-  constexpr paddr_t(segment_id_t seg, segment_off_t offset)
+  constexpr paddr_t(segment_id_t seg, seastore_off_t offset)
     : dev_addr((static_cast<common_addr_t>(seg.segment)
 	<< SEG_OFF_LEN_BITS) | static_cast<uint32_t>(offset)) {}
   constexpr paddr_t(common_addr_t val) : dev_addr(val) {}
+  constexpr paddr_t(block_id_t blk, seastore_off_t offset)
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:513]: block_id_t contains an offset, doesn't it?  Why isn't this device_id_t, seastore_off_t?

File: src/crimson/os/seastore/seastore_types.h [hunk 7]
```diff
@@ -468,6 +584,48 @@ struct paddr_t {
...
+    common_addr_t ret =
+      dev_addr &
+      ~static_cast<common_addr_t>(
+	std::numeric_limits<device_seastore_off_t>::max());
+    seastore_off_t l = static_cast<seastore_off_t>(dev_addr &
+      static_cast<common_addr_t>(
+	std::numeric_limits<device_seastore_off_t>::max()));
+    seastore_off_t r = static_cast<seastore_off_t>(o.dev_addr &
+      static_cast<common_addr_t>(
+	std::numeric_limits<device_seastore_off_t>::max()));
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_types.h:598]: This implementation is confusing.  o is being treated as though it is a blk_paddr_t, but relative addrs are (for the moment, at least) always seg_paddr_t.

**Comment by athanatos** [comment]
This is getting closer, but please make a pass and ensue that every type and constant added is necessary and as clear as possible.

**Comment by athanatos** [comment]
Ping me when it's ready for another review.  I'm on PTO until Jan 3rd, but I may have time before then.

File: src/crimson/os/seastore/seastore_types.h (hunk 3 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.h:178]: This is added to use 32 bit (unsigned) helpers such as ::max(). In other word, the purpose of this is to use entire 32 bit, which is the same size of the seastore_off_t,  excluding MSB.

File: src/crimson/os/seastore/seastore_types.h (hunk 4 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.h:477]: ok

File: src/crimson/os/seastore/seastore_types.h (hunk 5 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.h:478]: ok

File: src/crimson/os/seastore/seastore_types.h (hunk 6 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.h:513]: It seems that we need to discuss blk_paddr_t format. I added a comment as below.

**Comment by myoungwon** [comment]
> This patch seems to add several types:
> 
> device_block_id_t -- if this is meant to be similar to device_segment_id_t, it would be just a block offset. Is this just supposed to be an unsigned seastore_off_t? block_id_t -- combination of device_id_t and device_block_id_t as with segment_id_t? Is this actually useful? We want a segment_id_t because segment_cleaner internally deals with segments from multiple devices, but I don't see an obvious analogue for block devices that wouldn't just be a paddr_t. blk_paddr_t -- this makes sense
> 
> I don't think you actually need device_block_id_t or block_id_t.

There was a mismatch between us.
The reason why block_id_t is introduced here is that segment_id_t limits to represent entire address in the block device if segment_id_t is used as block id. 
Because current segment_id_t is 24 bits and seastore_off_t is 32 bits---if the block size is 4096 (block address 5000 -> block id 1 and offset 4), the limitation is only 64GB (4096 * max value of segment_id_t = 64GB). Therefore, at least, we need 20 bits to represents block offset here, which means that segment_id_t needs to be 36 bit and seastore_off_t (for block offset) is 20 bits.

How about the below formats for blk_paddr_t if we need to keep segment_id_t?
option 1) block address = segment_id_t * blocks_per_segment * block_size + block offset
option 2) block address = segment_id_t * blocks_per_segment + block id * block size + block offset
If block size is 4096, segment_id_t would be 0 and block offset would be 4096.

| device_id_t | segment_id_t |  blocks_per_segment (4 bits) | block offset |
<---------segment_id_t--------><----------------seastore_off_t--------------------->
<------------- 32 bit--------------><----------------32 bit-------------------------------->

or 

| device_id_t | segment_id_t |    block id  (4 bits)    | block offset (28 bits) |
<---------segment_id_t--------><----------------seastore_off_t--------------------->
<------------- 32 bit--------------><----------------32 bit-------------------------------->

?

**Comment by athanatos** [comment]
I'm catching up on reviews, I expect to get to this one early next week.

**Comment by myoungwon** [comment]
@athanatos Ping.

**Comment by athanatos** [comment]
> > This patch seems to add several types:
> > device_block_id_t -- if this is meant to be similar to device_segment_id_t, it would be just a block offset. Is this just supposed to be an unsigned seastore_off_t? block_id_t -- combination of device_id_t and device_block_id_t as with segment_id_t? Is this actually useful? We want a segment_id_t because segment_cleaner internally deals with segments from multiple devices, but I don't see an obvious analogue for block devices that wouldn't just be a paddr_t. blk_paddr_t -- this makes sense
> > I don't think you actually need device_block_id_t or block_id_t.
> 
> There was a mismatch between us. The reason why block_id_t is introduced here is that segment_id_t limits to represent entire address in the block device if segment_id_t is used as block id. Because current segment_id_t is 24 bits and seastore_off_t is 32 bits---if the block size is 4096 (block address 5000 -> block id 1 and offset 4), the limitation is only 64GB (4096 * max value of segment_id_t = 64GB). Therefore, at least, we need 20 bits to represents block offset here, which means that segment_id_t needs to be 36 bit and seastore_off_t (for block offset) is 20 bits.
> 
> Otherwise, how about the below formats for blk_paddr_t if we need to keep segment_id_t? option 1) block address = segment_id_t * blocks_per_segment * block_size + block offset.
> 
> | device_id_t | segment_id_t | blocks_per_segment (4 bits) | block offset | <---------segment_id_t--------><----------------seastore_off_t---------------------> <------------- 32 bit--------------><----------------32 bit--------------------------------> If block size is 4096, segment_id_t would be 0 and block offset would be 4096. or
> 
> option 2) block id = segment_id_t * blocks_per_segment (== max value of seg. block id (4bits)) + seg. block id block address = block id * block_size + block offset
> 
> | device_id_t | segment_id_t | seg. block id (4 bits) | block offset (28 bits) | <---------segment_id_t--------><----------------seastore_off_t---------------------> <------------- 32 bit--------------><----------------32 bit-------------------------------->
> 
> ?

I don't really understand the above, blk_paddr_t doesn't need a segment_id_t.  It sounds to me like the purpose for block_id_t is that seastore_off_t is only 32 bits.  In that case, introduce a 64 bit unsigned block_off_t and use that.  I still don't understand the purpose of the other types.

**Comment by myoungwon** [comment]
@athanatos Updated. Can you take a look?

**Review by athanatos — APPROVED** [review]
LGTM.  If you're ready, remove the WIP/DNM and I'll merge it tomorrow.

**Comment by myoungwon** [comment]
Removed.

**Pull request merged**